3.8/graphs-custom-pregel.md

```diff
@@ -0,0 +1,1550 @@
+["quote", 1, 2, ["foo", "bar"]] // evaluates to [1, 2, ["foo", "bar"]] -- unevaluated parameters
+```
+
+They are described in more detail below.
+
+The documentation refers to an array of length two as _pair_. The first entry is
+called `first` and the second entry `second`.
+
+### Truthiness of values
+
+A value is considered _false_ if it is boolean `false` or absent (`none`).
+All other values are considered true.
+
+### Special forms
+
+A _special form_ is special in the sense that it does not necessarily evaluate
+its parameters.
+
+#### `let` statement
+
+_binding values to variables_
+
+```js
+["let", [[name, value]...], expr...]
+```
+
+Expects as first parameter a list of name-value-pairs. Both members of each
+pair are evaluated. `first` has to evaluate to a string. Declared names become
+visible at the first expr.The following expressions are then evaluated in a
+context where the named variables are assigned to their given values. When
+evaluating the expression, `let` behaves like `seq`.
+
+[Variables](#variables) can be dereference using `var-ref`.
+
+```js
+> ["let", [["x", 12], ["y", 5]], ["+", ["var-ref", "x"], ["var-ref", "y"]]]
+ = 17
+```
+
+#### `seq` statement
+
+_sequence of commands_
+
+```js
+["seq", expr ...]
+```
+
+`seq` evaluates `expr` in order. The result value is the result value of the
+last expression. An empty `seq` evaluates to `none`.
+
+```js
+> ["seq", ["print", "Hello World!"], 2, 3]
+Hello World!
+ = 3
+```
+
+#### `if` statement
+
+_classical if-elseif-else-statement_
+
+```js
+["if", [cond, body], ...]
+```
+
+Takes pairs `[cond, body]` of conditions `cond` and expression `body` and
+evaluates the first `body` for which `cond` evaluates to a value that is
+considered true. It does not evaluate the other `cond`s. If no condition
+matches, it evaluates to `none`. To simulate an `else` statement, set the
+last condition to `true`.
+
+```js
+> ["if", [
+        ["lt?", ["var-ref", "x"], 0],
+        ["-", 0, ["var-ref", "x"]]
+    ], [
+        true, // else
+        ["var-ref", "x"]
+    ]]
+ = 5
+```
+
+#### `match` statement
+
+_not-so-classical switch-statement_
+
+```js
+["match", proto, [c, body]...]
+```
+
+First evaluates `proto`, then evaluates each `c` until `["eq?", val, c]` is
+considered true. Then the corresponding `body` is evaluated and its return
+value is returned. If no branch matches, `none` is returned. This is a C-like
+`switch` statement except that its `case`-values are not treated as constants.
+
+```js
+> ["match", 5,
+  [1, ["A"]],
+  [2, ["B"]],
+  [3, ["C"]],
+  [4, ["D"]],
+  [5, ["E"]]
+]
+  = "E"
+```
+
+#### `for-each` statement
+
+```js
+["for-each", [[var, list]...] expr...]
+```
+
+Behaves similar to `let` but expects a list as `value` for each variable.
+It then produces the cartesian product of all lists and evaluates its
+expression for each n-tuple. The return value is always `none`. The order
+is guaranteed to be reversed lexicographic order.
```

Comment:
Would it be sensible to add a sentence what happens when the list of `[var, list]` pairs is empty?

Reply:
Well, if it would actually produce the Cartesian product of all lists and evaluates its expression for each n-tuple, then it would have to do one iteration. 😆 Do we want that?

Reply:
Well, it produces the Cartesian product of all lists and evaluates its expression for each n-tuple

Reply:
True :) But I think it helps the reader to explicitly write down that it is allowed at all, and that the expression will be evaluated exactly once in that case, because it's probably not self-evident for everyone - and even for those who expect it might not be confident about the actual implementation in such an edge case from the documentation alone.